src/Combinatorics/EnumerativeCombinatorics/tableaux.jl

```diff
@@ -492,109 +492,199 @@ end
+  s = shape(S)
+  weight = Oscar.weight(S)
+
+  if isnothing(state)
+    state = SemiStandardTableauxFixedShapeAndWeightState{T}()
+    if isempty(s)
+      state.n = 0
+      return young_tableau(Vector{T}[], check = false), state
+    end
+
+    state.n = 1
+    state.increaseN = true
+    state.tab = young_tableau([T[0 for j = 1:s[i]] for i = 1:length(s)], check = false)
+    state.boxes_filled = zeros(Int, length(s))
+    state.n_used_weight = zeros(Int, length(weight))
+    state.row_pointer = zeros(Int, length(weight), maximum(weight))
   end
-  ls = length(s)
 
-  tab = young_tableau([ T[0 for j = 1:s[i]] for i = 1:length(s)], check = false)
-  sub_s = zeros(Int, length(s))
+  # the integer that we currently try to write into the tableau
+  n = state.n
 
-  #tracker_row = zeros(Integer, n_max)
+  is_zero(n) && return nothing
 
-  function rec_sst!(n::Int)
+  # whether we should increase or decrease n in the next iteration of the while loop
+  increaseN = state.increaseN
+  # the tableau filled in-place
+  tab = state.tab
+  # boxes_filled[i] is the number of boxes we successfully filled in row i
+  boxes_filled = state.boxes_filled
+
+  while n > 0
+    if is_zero(weight[n])
+      if increaseN
+        n += 1
+      else
+        n -= 1
+      end
+      continue
+    end
 
-    #fill the remaining boxes if possible, else set them to 0
+    # n is the last possible entry, so we try to fill the remaining boxes with n
     if n == length(weight)
-      for i = 1:ls
-        for j = sub_s[i] + 1:s[i]
-          tab[i][j] = T(n)
-          if i != 1 && tab[i - 1][j] == n
-            for k = 1:i
-              for l = sub_s[k] + 1:s[k]
-                tab[i][j] = T(0)
-              end
-            end
-            return
-          end
-        end
+      goodTableau = _is_good_tableau!(n, S, state)
+      n -= 1
+      increaseN = false
+      if goodTableau
+        # we discovered a tableau
+        state.n = n
+        state.increaseN = increaseN
+        return young_tableau([copy(row) for row in tab], check = false), state
       end
-      push!(tabs, young_tableau([copy(row) for row in tab], check = false))
+      # we did not discover a tableau and have to go back to the step n - 1
+      continue
+    end
 
-      return
+    # fill n into the boxes (modifies state in-place)
+    _fill_boxes!(n, S, state)
+    n = state.n
+    increaseN = state.increaseN
+  end
 
-      #skip to next step if weight[n] == 0
-    elseif weight[n] == 0
-      rec_sst!(n + 1)
-      return
-    end
+  return nothing
+end
 
-    #here starts the main part of the function
-    tracker_row = zeros(Int, weight[n])
+function _fill_boxes!(n::Int, S::SemiStandardTableauxFixedShapeAndWeight{T}, state::SemiStandardTableauxFixedShapeAndWeightState{T}) where T
+  # Note: n might be different from state.n if the calling function did not update
+  # state yet
```

Comment:
Is there a way to work around this, maybe just ensure the iterator updates the state before calling this? I'm not sure this really matters though.

Reply:
Yes, you're right. I changed so that `state.n` is updated before this function is called. (I still think it's clearer if `n` is an explicit argument of the function, so I left this as it was.)